File: backend/app/api/endpoints/asset.py
import polars as pl
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app import crud
from app.crud import update_watchlist_item_alert_percentage
from app.database import get_db
from app.logger import logger
from app.schemas.asset import AssetInWatchlist, AssetListSchema, AssetSchema
from app.services.asset_service import generate_asset_list

router = APIRouter(prefix="/asset", tags=["asset"])


@router.get("/", response_model=list[AssetSchema])
def get_all_assets(db: Session = Depends(get_db)):
    return crud.asset.get_all_assets(db)


@router.get("/asset_list", response_model=list[AssetListSchema])
def get_asset_list(db: Session = Depends(get_db)):
    assets = crud.asset.get_all_assets(db)
    latest_timeseries = crud.timeseries.get_latest_price_and_changes(db)
    asset_list = generate_asset_list(assets, latest_timeseries)

    return asset_list


@router.get("/get_asset_by_ticker", response_model=AssetListSchema)
def get_asset_by_ticker(ticker: str, db: Session = Depends(get_db)):
    asset = crud.asset.get_asset_by_ticker(db, ticker)

    if asset is None:
        logger.warning(f"Asset not found: ticker={ticker}")
        raise HTTPException(status_code=404, detail=f"Asset {ticker} not found")

    latest_timeseries = crud.timeseries.get_latest_price_and_changes(db)
    filtered = latest_timeseries.filter(pl.col("asset_id") == asset["id"])

    if len(filtered) == 0:
        logger.warning(
            f"No price data available: ticker={ticker}, asset_id={asset['id']}"
        )

    ts_data = filtered.to_dicts()[0] if len(filtered) > 0 else {}
    asset.update(ts_data)

    return asset


@router.get("/check_asset_in_watchlist", response_model=AssetInWatchlist)
def check_asset_in_watchlist(
    ticker: str, user_id: str, db: Session = Depends(get_db)
) -> AssetInWatchlist | None:
    asset = crud.asset.get_asset_by_ticker(db, ticker)
    user_watchlist_items = crud.get_watchlist_items(user_id, db)
    watchlist_ids = [item.asset_id for item in user_watchlist_items]
    percentage_map = {
        item.asset_id: item.alert_percentage_change for item in user_watchlist_items
    }
    alert_percentage = percentage_map.get(asset["id"], None)

    return AssetInWatchlist(
        asset_in_watchlist=asset["id"] in watchlist_ids,
        alert_percentage=alert_percentage,
    )


@router.post("/watchlist_alerts")
def watchlist_alerts(
    asset_id: int,
    user_id: str,
    enable_price_alerts: bool,
    asset_alert_percentage: int,
    db: Session = Depends(get_db),
):
    user_watchlist_items = crud.get_watchlist_items(user_id, db)
    watchlist_ids = [item.asset_id for item in user_watchlist_items]

    if asset_id not in watchlist_ids:
        logger.warning(
            f"Alert config failed - asset not in watchlist: asset_id={asset_id}"
        )
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Asset not found in watchlist"
        )

    result = update_watchlist_item_alert_percentage(
        asset_id, user_id, enable_price_alerts, asset_alert_percentage, db
    )

    if not result:
        logger.error(f"Failed to update watchlist alert: asset_id={asset_id}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Watchlist item not found"
        )

    if enable_price_alerts:
        logger.info(
            f"Watchlist alert enabled: asset_id={asset_id}, user=...{user_id[-8:]}, threshold={asset_alert_percentage}%"
        )
    else:
        logger.info(
            f"Watchlist alert disabled: asset_id={asset_id}, user=...{user_id[-8:]}"
        )

    return {
        "message": "Alert settings updated successfully",
        "alert_percentage": result.alert_percentage_change,
    }

File: dashboard/src/api/asset.ts
import React from "react";
import { toast } from "sonner";
import { apiClient, ApiError } from "@/lib/api-client";
import { Asset, Portfolio } from "@/types/custom-types";
import { format } from "date-fns";

export function useTransactionType() {
  return React.useState<"buy" | "sell">("buy");
}

export const getAssetList = async () => {
  try {
    return await apiClient.get<Asset[]>("/asset/asset_list");
  } catch (error) {
    const apiError = error as ApiError;

    console.error("Asset list fetch failed:", {
      status: apiError.status,
      error: apiError.message,
    });

    throw apiError;
  }
};

export const getTimeseriesDataForAsset = async (
  assetId: number,
  setTimeseries: (data: Portfolio[]) => void,
  timeseriesRange: string,
) => {
  try {
    const data = await apiClient.get<Portfolio[]>(
      "/timeseries/timeseries_for_asset",
      {
        params: {
          asset_id: assetId,
          timeseries_range: timeseriesRange,
        },
      },
    );
    setTimeseries(data);
  } catch (error) {
    const apiError = error as ApiError;
    console.error("Error fetching asset data:", apiError);
    toast("There was an error fetching asset data.");
  }
};

export const getAssetByTicker = async (ticker: string | undefined) => {
  if (!ticker) return;

  try {
    return await apiClient.get<Asset>("/asset/get_asset_by_ticker", {
      params: { ticker },
    });
  } catch (error) {
    const apiError = error as ApiError;
    console.error("Error fetching asset data:", apiError);
    toast("There was an error fetching asset data.");
  }
};

export interface AssetInWatchlist {
  asset_in_watchlist: boolean;
  alert_percentage: number;
}

export const checkAssetInWatchlist = async (
  ticker: string | undefined,
  user_id: string | null,
): Promise<AssetInWatchlist | undefined> => {
  if (!ticker || !user_id) return undefined;

  try {
    return await apiClient.get<AssetInWatchlist>(
      "/asset/check_asset_in_watchlist",
      {
        params: {
          ticker,
          user_id,
        },
      },
    );
  } catch (error) {
    const apiError = error as ApiError;
    console.error("Error fetching asset data:", apiError);
    toast("There was an error fetching asset data.");
  }
};

export const saveAlertsChange = async (
  assetId: number | undefined,
  user_id: string | null,
  enablePriceAlerts: boolean,
  assetAlertPercentage: number,
) => {
  if (assetId === undefined || user_id === null) {
    toast.error("An error occurred.");
    return;
  }

  try {
    await apiClient.post("/asset/watchlist_alerts", null, {
      params: {
        asset_id: assetId,
        user_id: user_id,
        enable_price_alerts: enablePriceAlerts,
        asset_alert_percentage: assetAlertPercentage,
      },
    });

    toast.success("Alert settings updated successfully");
  } catch (error) {
    console.error("Failed to update alert settings:", error);
    toast.error("Failed to update alert settings");
    throw error;
  }
};

export const getPriceOnDate = async (date: Date, asset_id: number | null) => {
  const stringDate = format(date, "yyyy-MM-dd");
  try {
    return await apiClient.get<number>(
      `/asset/price_on_date?asset_id=${asset_id}&date=${stringDate}`,
    );
  } catch (error) {
    const apiError = error as ApiError;
    console.error("Error fetching asset price on date:", apiError);
    toast("There was an error fetching asset price on date.");
  }
};
